core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ZooCache.java

```diff
@@ -152,53 +184,78 @@ public void process(WatchedEvent event) {
+    this.zk = requireNonNull(zk);
+    this.cache = Caches.getInstance().createNewBuilder(Caches.CacheName.ZOO_CACHE, false)
+        .ticker(requireNonNull(ticker)).expireAfterAccess(CACHE_DURATION).build();
+    this.nodeCache = cache.asMap();
+    this.pathsToWatch = requireNonNull(pathsToWatch);
+    setupWatchers(pathsToWatch);
+    log.trace("{} created new cache", cacheId, new Exception());
   }
 
-  public ZooCache(ZooSession zk, Optional<ZooCacheWatcher> watcher, Duration timeout) {
-    this.zk = requireNonNull(zk);
-    this.externalWatcher = watcher;
-    RemovalListener<String,ZcNode> removalListerner = (path, zcNode, reason) -> {
-      try {
-        log.trace("{} removing watches for {} because {} accesses {}", cacheId, path, reason,
-            zcNode == null ? -1 : zcNode.getAccessCount());
-        zk.removeWatches(path, ZooCache.this.watcher, Watcher.WatcherType.Any, false);
-      } catch (InterruptedException | KeeperException | RuntimeException e) {
-        log.warn("{} failed to remove watches on path {} in zookeeper", cacheId, path, e);
+  public void addZooCacheWatcher(ZooCacheWatcher watcher) {
+    externalWatchers.add(requireNonNull(watcher));
+  }
+
+  // Visible for testing
+  protected void setupWatchers(Set<String> pathsToWatch) {
+
+    for (String left : pathsToWatch) {
+      for (String right : pathsToWatch) {
+        if (!left.equals(right) && left.contains(right)) {
+          throw new IllegalArgumentException(
+              "Overlapping paths found in paths to watch. left: " + left + ", right: " + right);
+        }
       }
-    };
-    // Must register the removal listener using evictionListener inorder for removal to be mutually
-    // exclusive with any other operations on the same path. This is important for watcher
-    // consistency, concurrently adding and removing watches for the same path would leave zoocache
-    // in a really bad state. The cache builder has another way to register a removal listener that
-    // is not mutually exclusive.
-    Cache<String,ZcNode> cache =
-        Caches.getInstance().createNewBuilder(Caches.CacheName.ZOO_CACHE, false)
-            .expireAfterAccess(timeout).evictionListener(removalListerner).build();
-    nodeCache = cache.asMap();
-    log.trace("{} created new cache", cacheId, new Exception());
+    }
+
+    try {
+      for (String path : pathsToWatch) {
+        watchedPaths.add(path);
+        zk.addPersistentRecursiveWatcher(path, this.watcher);
+        log.trace("Added persistent recursive watcher at {}", path);
+      }
+    } catch (KeeperException | InterruptedException e) {
+      throw new RuntimeException("Error setting up persistent recursive watcher", e);
+    }
+  }
+
+  private boolean isWatchedPath(String path) {
+    // Check that the path is equal to, or a descendant of, a watched path
+    for (String watchedPath : watchedPaths) {
+      if (path.startsWith(watchedPath)) {
+        return true;
+      }
+    }
+    return false;
```

Comment:
Can use the floor function on treeset to quickly find the string with the same prefix.  I was experimenting with this and found that calling startsWith is not sufficient to ensure something is watched.  For example if the code watches the path `/accumulo/iid/gc/lock`, then the path `/accumulo/iid/gc/lock1` starts with `/accumulo/iid/gc/lock` but it is not watched because `lock1` is a sibling to `lock` in ZK.  Added `floor.equals(path) ||  path.startsWith(floor+"/")` to deal with this, but I don't like that it allocates a string.

```suggestion
  var floor = ts.floor(path);
  return floor != null && (floor.equals(path) ||  path.startsWith(floor+"/"));
```